### PR TITLE
It only a extraction of the original. Must be reworked later

### PR DESCRIPTION
**Pull request by TheMeinerLP**
I have only copy the code and do in extra maven modules for more overview and legibility. :)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    stage('Release ZIP') {
+      steps {
+        sh '''mkdir -p temp;
+        cp MobModule/target/*.jar temp/;
+        cp MobPlugin/target/*.jar temp/;'''
+        zip archive: true, dir: 'temp', glob: '', zipFile: 'MobSystem.zip'
+        sh 'rm -r temp/';
+      }
+    }
```
Comment (GiantTreeLP) [inline Jenkinsfile:40]: Creating a ZIP file is unnecessary, when it is not archived.

File: MobCore/src/main/java/eu/cloudnetservice/cloudmob/MobItemLayout.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.List;
+
+public final class MobItemLayout implements Cloneable {
+
+    /**
+     * itemIds are not supported in all versions, use {@link MobItemLayout#itemName} instead
+     */
+
+    @Deprecated
+    private int itemId;
```
Comment (GiantTreeLP) [inline MobCore/src/main/java/eu/cloudnetservice/cloudmob/MobItemLayout.java]: We should drop this, as this version targets Spigot 1.13+

File: MobCore/src/main/java/eu/cloudnetservice/cloudmob/ServerMob.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    protected UUID uniqueId;
+    protected String display;
+    protected String name;
+    protected String type;
+    protected String targetGroup;
+    /**
+     * itemIds are not supported in all versions, use {@link ServerMob#itemName} instead
+     */
+    @Deprecated
+    protected Integer itemId;
```
Comment (GiantTreeLP) [inline MobCore/src/main/java/eu/cloudnetservice/cloudmob/ServerMob.java]: See the above comment

File: MobModule/src/main/java/eu/cloudservice/cloudmob/MobDatabase.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        for (ServerMob serverMob : mobMap.values()) {
+            if (serverMob.getItemId() == null) {
+                serverMob.setItemId(138);
+                injectable = true;
+            }
+            if (serverMob.getAutoJoin() == null) {
+                serverMob.setAutoJoin(false);
+                injectable = true;
+            }
+        }
```
Comment (GiantTreeLP) [inline MobModule/src/main/java/eu/cloudservice/cloudmob/MobDatabase.java]: Was this necessary for an update?

File: MobModule/src/main/java/eu/cloudservice/cloudmob/MobModule.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import de.dytanic.cloudnet.lib.server.ServerGroupMode;
+import de.dytanic.cloudnetcore.api.CoreModule;
+import de.dytanic.cloudnetcore.api.event.network.ChannelInitEvent;
+import de.dytanic.cloudnetcore.api.event.network.UpdateAllEvent;
+import de.dytanic.cloudnetcore.network.components.MinecraftServer;
+import eu.cloudservice.cloudmob.in.PacketInAddMob;
+import eu.cloudservice.cloudmob.in.PacketInRemoveMob;
+import eu.cloudservice.cloudmob.out.PacketOutMobSelector;
+
+public class MobModule extends CoreModule implements IEventListener<UpdateAllEvent> {
```
Comment (GiantTreeLP) [inline MobModule/src/main/java/eu/cloudservice/cloudmob/MobModule.java]: Having this class implement an EventListener seems like an architectural mistake

File: MobModule/src/main/java/eu/cloudservice/cloudmob/MobModule.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            if (event.getINetworkComponent() instanceof MinecraftServer) {
+                MinecraftServer minecraftServer = (MinecraftServer) event.getINetworkComponent();
+
+                if (minecraftServer.getGroupMode().equals(ServerGroupMode.LOBBY) || minecraftServer.getGroupMode()
+                                                                                                   .equals(ServerGroupMode.STATIC_LOBBY)) {
+                    minecraftServer.sendPacket(new PacketOutMobSelector(configMobs.load(), mobDatabase.loadAll()));
+                }
+            }
+        }
+    }
```
Comment (GiantTreeLP) [inline MobModule/src/main/java/eu/cloudservice/cloudmob/MobModule.java]: This should be a separate class

File: MobModule/src/main/resources/module.properties [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) Tarek Hosni El Alaoui 2017
+#
```
Comment (GiantTreeLP) [inline MobModule/src/main/resources/module.properties]: You may drop this bogus "Copyright"

File: MobModule/src/main/java/eu/cloudservice/cloudmob/in/PacketInRemoveMob.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import eu.cloudnetservice.cloudmob.ServerMob;
+import eu.cloudservice.cloudmob.MobModule;
+
+public class PacketInRemoveMob extends PacketInHandler {
+
+    public void handleInput(Document packet, PacketSender packetSender) {
+        ServerMob serverMob = packet.getObject("mob", ServerMob.TYPE);
+        MobModule.getInstance().getMobDatabase().remove(serverMob);
+        CloudNet.getInstance().getNetworkManager().reload();
+        CloudNet.getInstance().getNetworkManager().updateAll();
```
Comment (GiantTreeLP) [inline MobModule/src/main/java/eu/cloudservice/cloudmob/in/PacketInRemoveMob.java]: This does not look right

File: MobModule/src/main/java/eu/cloudservice/cloudmob/in/PacketInAddMob.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+import eu.cloudnetservice.cloudmob.ServerMob;
+import eu.cloudservice.cloudmob.MobModule;
+
+public class PacketInAddMob extends PacketInHandler {
+
+    public void handleInput(Document packet, PacketSender packetSender) {
+        ServerMob serverMob = packet.getObject("mob", ServerMob.TYPE);
+        MobModule.getInstance().getMobDatabase().append(serverMob);
+        CloudNet.getInstance().getNetworkManager().reload();
+        CloudNet.getInstance().getNetworkManager().updateAll();
```
Comment (GiantTreeLP) [inline MobModule/src/main/java/eu/cloudservice/cloudmob/in/PacketInAddMob.java]: This seems like bad design

File: MobModule/pom.xml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>MobSystem</artifactId>
+        <groupId>eu.cloudnetservice</groupId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>MobModule</artifactId>
```
Comment (GiantTreeLP) [inline MobModule/pom.xml]: As others have noticed, we should stick to lower-case artifact IDs

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/MobSelector.java [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+            location.getZ(),
+            location.getYaw(),
+            location.getPitch());
+    }
+
+    public Mob spawnMob(final MobConfig mobConfig, final UUID uuid, final ServerMob serverMob) {
+        Location location = toLocation(serverMob.getPosition());
+
+        if (!location.getChunk().isLoaded()) {
+            location.getChunk().load();
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/MobSelector.java]: We should not spawn mobs, when the given chunk is unloaded.
Delay that process to when the chunk is actually loaded and cleanup upon unloading

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/listener/v18_112/ArmorStandListener.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    @EventHandler
+    public void handle(PlayerArmorStandManipulateEvent e) {
+        Mob mob = MobSelector.getInstance().getMobs().get(e.getRightClicked().getUniqueId());
+        if (mob != null) {
+            e.setCancelled(true);
+        }
+    }
+
+    @EventHandler
+    public void handle(ItemDespawnEvent e) {
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/listener/v18_112/ArmorStandListener.java]: How come the item would despawn?

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /**
+     * @deprecated will only work in versions lower than 1.13
+     */
+    @Deprecated
+    public ItemStackBuilder(int material) {
+        this.itemStack = new ItemStack(Material.getMaterial(material));
+        this.itemMeta = itemStack.getItemMeta();
+    }
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java]: Remove this (see earlier comments)

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+    /**
+     * @deprecated will only work in versions lower than 1.13
+     */
+    @Deprecated
+    public ItemStackBuilder(int material, int amount) {
+        this.itemStack = new ItemStack(material, amount);
+        this.itemMeta = itemStack.getItemMeta();
+    }
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java]: Remove this (see earlier comments)

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+    /**
+     * @deprecated will only work in versions lower than 1.13
+     */
+    @Deprecated
+    public ItemStackBuilder(int material, int amount, int sub) {
+        this.itemStack = new ItemStack(material, amount, (short) sub);
+        this.itemMeta = itemStack.getItemMeta();
+    }
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java]: Remove this (see earlier comments)

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+                    "getMaterialIgnoreVersion",
+                    String.format(
+                        "Can't get material by id %d! Beginning with MC 1.13 you HAVE to use material names!%n",
+                        id),
+                    exception);
+                return null;
+            }
+        }
+        return Material.getMaterial(name);
+    }
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java]: Remove this (see earlier comments)

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        return new ItemStackBuilder(material, amount);
+    }
+
+    /**
+     * @deprecated will only work in versions lower than 1.13
+     */
+    @Deprecated
+    public static ItemStackBuilder builder(int material, int amount, int sub) {
+        return new ItemStackBuilder(material, amount, sub);
+    }
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ItemStackBuilder.java]: Remove this (see earlier comments)

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ReflectionUtil.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+import eu.cloudnetservice.cloudmob.ServerMob;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.entity.EntityType;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Wither;
+
+public final class ReflectionUtil {
```
Comment (GiantTreeLP) [inline MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ReflectionUtil.java]: I don't think this would be needed anymore

File: readme.md [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+
+<dependencies>
+    <!-- RootProject -->
+    <dependency>
+        <groupId>eu.cloudnetservice</groupId>
+        <artifactId>MobSystem</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <scope>provided</scope>
+    </dependency>
+</dependencies>
```
Comment (GiantTreeLP) [inline readme.md]: You can't (and should not) depend on the root project, that contains no code

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

File: MobModule/src/main/java/eu/cloudservice/cloudmob/MobDatabase.java (hunk 4 above)
Comment (TheMeinerLP) [reply MobModule/src/main/java/eu/cloudservice/cloudmob/MobDatabase.java]: Its the old code from Tarek we must be rewoking later...

File: MobModule/pom.xml (hunk 10 above)
Comment (TheMeinerLP) [reply MobModule/pom.xml]: okay, i renaming

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ReflectionUtil.java (hunk 18 above)
Comment (TheMeinerLP) [reply MobPlugin/src/main/java/eu/cloudservice/cloudmob/util/ReflectionUtil.java]: We needted for armorstand

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/listener/v18_112/ArmorStandListener.java (hunk 12 above)
Comment (TheMeinerLP) [reply MobPlugin/src/main/java/eu/cloudservice/cloudmob/listener/v18_112/ArmorStandListener.java]: Idk, o have only copied this code.

File: MobCore/src/main/java/eu/cloudnetservice/cloudmob/MobItemLayout.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
+package eu.cloudnetservice.cloudmob;
+
+import java.util.List;
+
+public final class MobItemLayout implements Cloneable {
+
+    /**
+     * itemIds are not supported in all versions, use {@link MobItemLayout#itemName} instead
+     */
```
Comment (GiantTreeLP) [inline MobCore/src/main/java/eu/cloudnetservice/cloudmob/MobItemLayout.java]: Remove this comment as well

File: MobModule/pom.xml (hunk 10 above)
Comment (GiantTreeLP) [reply MobModule/pom.xml]: Please be consistent

File: MobModule/pom.xml [hunk 21]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>mob-system</artifactId>
+        <groupId>eu.cloudnetservice</groupId>
```
Comment (GiantTreeLP) [inline MobModule/pom.xml:7]: When we are already pedantic about naming, we should make sure to not restrict ourselves to just the base group id.
Maybe something like `eu.cloudnetservice.v2.mobs` would be appropriate

File: MobPlugin/pom.xml [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+                    <relocations>
+                        <relocation>
+                            <!-- Fixes issues with missing TypeToken.getParameterized method -->
+                            <pattern>com.google.gson.reflect.TypeToken</pattern>
+                            <shadedPattern>eu.cloudnetservice.relocation.gson.reflect.TypeToken</shadedPattern>
+                        </relocation>
+                        <relocation>
+                            <!-- Fixes issues with deprecated JsonParser methods -->
+                            <pattern>com.google.gson.JsonParser</pattern>
+                            <shadedPattern>eu.cloudnetservice.relocation.gson.JsonParser</shadedPattern>
```
Comment (GiantTreeLP) [inline MobPlugin/pom.xml]: These can be condensed into one relocation pattern, if wanted

File: MobPlugin/src/main/java/eu/cloudservice/cloudmob/listener/v18_112/ArmorStandListener.java (hunk 12 above)
Comment (GiantTreeLP) [reply MobPlugin/src/main/java/eu/cloudservice/cloudmob/listener/v18_112/ArmorStandListener.java]: This was a question, not a command to remove this code.
In practice, the item should be spawned with the means necessary for it to never despawn

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]
Please also reformat the code, manually if necessary

File: MobCore/src/main/java/eu/cloudnetservice/v2/mobs/ServerMob.java [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+
+    public String getTargetGroup() {
+        return targetGroup;
+    }
+
+    public void setTargetGroup(String targetGroup) {
+        this.targetGroup = targetGroup;
+    }
+
+    public Boolean getAutoJoin() {
```
Comment (GiantTreeLP) [inline MobCore/src/main/java/eu/cloudnetservice/v2/mobs/ServerMob.java]: This should probably be a primitive

File: MobCore/src/main/java/eu/cloudnetservice/v2/mobs/ServerMob.java [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+        return autoJoin;
+    }
+
+    public void setAutoJoin(Boolean autoJoin) {
```
Comment (GiantTreeLP) [inline MobCore/src/main/java/eu/cloudnetservice/v2/mobs/ServerMob.java]: This should probably be a primitive

File: MobPlugin/pom.xml [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+                            <!- Fixes issues with missing TypeToken.getParameterized method
+                            <pattern>com.google.gson.reflect.TypeToken</pattern>
+                            <shadedPattern>eu.cloudnetservice.relocation.gson.reflect.TypeToken</shadedPattern>
+                        </relocation>
+                        <relocation>
+                            <!- Fixes issues with deprecated JsonParser methods
+                            <pattern>com.google.gson.JsonParser</pattern>
+                            <shadedPattern>eu.cloudnetservice.relocation.gson.JsonParser</shadedPattern>
+                        </relocation>
+                    </relocations>-->
```
Comment (GiantTreeLP) [inline MobPlugin/pom.xml:100]: Is this not needed?

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]
The package names and the groupId mismatch

File: MobPlugin/pom.xml (hunk 25 above)
Comment (TheMeinerLP) [reply MobPlugin/pom.xml:100]: I think not.

**Review by GiantTreeLP — APPROVED** [review]

**Comment by GiantTreeLP** [comment]
For a first version, this is okay
We should definitely think about unifying the group ids and package names

**Pull request merged**